pkg/controller/service/servicecontroller.go

```diff
@@ -767,3 +737,64 @@ func (s *cachedService) nextRetryDelay() time.Duration {
 func (s *cachedService) resetRetryDelay() {
 	s.lastRetryDelay = time.Duration(0)
 }
+
+
+// syncService will sync the Service with the given key if it has had its expectations fulfilled,
+// meaning it did not expect to see any more of its pods created or deleted. This function is not meant to be
+// invoked concurrently with the same key.
+func (sc *ServiceController) syncService(key string) error {
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished syncing service %q (%v)", key, time.Now().Sub(startTime))
+	}()
+	//obj shows the latest service info from apiserver
+	obj, exists, err := sc.serviceStore.Store.GetByKey(key)
+	if !exists {
+		glog.Infof("Service has been deleted %v", key)
+		return nil
+	}
+	if err != nil {
+		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
+		sc.queue.Add(key)
+		return err
+	}
+
+	deltaService, ok := obj.(*api.Service)
+
+	var namespacedName types.NamespacedName
+	var cachedService *cachedService
+	if !ok {
+		key, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return fmt.Errorf("object contained wasn't a service or a deleted key: %+v", obj)
+		}
+		cachedService, ok = sc.cache.get(key.Key)
+		if !ok {
+			return fmt.Errorf("Service %s not in cache even though the watcher thought it was. Ignoring the deletion.", key)
+		}
+		deltaService = cachedService.lastState
+		obj = deltaService
+		namespacedName = types.NamespacedName{Namespace: deltaService.Namespace, Name: deltaService.Name}
+	} else {
+		namespacedName.Namespace = deltaService.Namespace
+		namespacedName.Name = deltaService.Name
+		cachedService = sc.cache.getOrCreate(namespacedName.String())
+	}
+
+	err, retryDelay := sc.processDelta(cachedService, deltaService, namespacedName)
```

Comment:
I just kept the old variable names, the logic is not actually processing the delta. the service store hold the latest service info, and on service updated, we simply add the service into queue and check if it needs update in createLoadBalancerIfNeeded(). I will change the variable and function name to avoid confusion, and add check to controller like 
```
UpdateFunc: func(old, cur interface{}) {
				if sc.needsUpdate(old, cur) {
					sc.enqueueService(cur)
				}
			},
```